### ISSUE TITLE
Consider using netcdf4 groups in SWALS to combine multidomain outputs into a single file

### ISSUE THREAD

**Issue by gareth-d-ga** [GeoscienceAustralia/ptha#10]
Currently SWALS writes multiple output files for each domain:
* A netcdf with gridded outputs
* A netcdf with gauge outputs
* Some text metadata

For large multidomains with hundreds of domains, this can lead to > 1000 files for a single model run. 
* This can be a problem, e.g., if running hundreds of scenarios then we start approaching file-count limits on some supercomputers (e.g. NCI). 

An alternative is to use netcdf groups to combine gauge/grid outputs from each domain in a single netcdf file. 
* This would require updates to SWALS including the post-processing scripts. 

A potential downside is that netcdf-groups might not play nicely with some other tools (like ncview). So ideally the netcdf-group format could be implemented as a compile-time option, with post-processing routines seamlessly working with either format.

**Comment by gareth-d-ga** [comment]
While not related to netcdf directly, pull request #22 has reduced the number of output files.

**Comment by gareth-d-ga** [comment]
While not exactly matching the point above, this can be done in a post-processing step using e.g. 
```
ncecat --gag */Grid*.nc -o test.nc
```
from inside a multidomain folder. 

The resulting file can be viewed with ncview, and read with R's ncdf4 package. 

It would be better to improve the group names (to include the domain folder). 

The above is at least useful for testing, or, could be an alternative way to manage the file counts. We could also introduce some compression, e.g.
 ```
ncecat -4 --deflate 9 --gag */Grid*.nc -o test3.nc
```

**Comment by gareth-d-ga** [comment]
`ncrename` might be able to change groups names

**Comment by gareth-d-ga** [comment]
This version iteratively adds netcdf files into groups, while using group names to reflect the directory structure. Run from the multidomain directory.
```
for i in RUN*/Grid*.nc; do echo $i; ncecat -4 --deflate 4 --no_tmp_fl -A -M -G $i --gag $i -o test7.nc; done
```
We can also put grids and gauges in the one file
```
for i in RUN*/G*.nc; do echo $i; ncecat -4 --deflate 4 --no_tmp_fl -A -M -G $i --gag $i -o test7.nc; done
```

**Comment by gareth-d-ga** [comment]
This looks promising to make the group names correct
```
for i in RUN*/Grid*.nc; do
    ncks -4 -A -G "$i" "$i" test7.nc
done
```

**Comment by gareth-d-ga** [comment]
The previous command became slow with many files. The following seems more efficient (edited from Google AI Mode).
```
# 1. Setup temporary workspace
mkdir -p ./tmp_groups
rm -f test7.nc

# 2. Step One: Create staged files (Metadata + Data copy)
# This part can be done in parallel.
echo "Step 1: Staging files into group structures..."
for i in RUN*/Grid*.nc; do
    # Create a safe temp filename
    tmp_file="./tmp_groups/${i//\//_}"
    grp_path=$i 
    # "/$(dirname "$i")/$(basename "$i" .nc)/"
    
    # Place data into the correct group in a separate file
    ncks -4 -G "$grp_path" "$i" "$tmp_file"
    
    # Control parallelism (adjust '4' to your number of CPU cores)
    #if [[ $(jobs -r | wc -l) -ge 4 ]]; then wait; fi
done
wait

# 3. Step Two: Efficiently merge into the final file
# We use ncks -A because it respects the internal group paths of the source
echo "Step 2: Performing final merge..."
staged_files=(./tmp_groups/*.nc)

for tmp in "${staged_files[@]}"; do
    echo "Merging $tmp..."
    # --no_tmp_fl prevents creating a copy of test7.nc every time
    ncks -4 -A --no_tmp_fl "$tmp" test7.nc
done

# 4. Cleanup
#rm -rf ./tmp_groups
echo "Done! Output saved to test7.nc"
```

**Comment by gareth-d-ga** [comment]
Here storing gauges and grids in the same file -- it would be good to also get the text metadata in there
```
# 1. Setup temporary workspace
mkdir -p ./tmp_groups
rm test7.nc

# 2. Step One: Create staged files (Metadata + Data copy)
# This part can be done in parallel.
echo "Step 1: Staging files into group structures..."
for i in RUN*/*.nc; do
    # Create a safe temp filename
    tmp_file="./tmp_groups/${i//\//_}"
    
    # Place data into the correct group in a separate file
    ncks -4 -G "$i" "$i" "$tmp_file"
    
    # Control parallelism (adjust '4' to your number of CPU cores)
    #if [[ $(jobs -r | wc -l) -ge 4 ]]; then wait; fi
done
wait

# 3. Step Two: Efficiently merge into the final file
# We use ncks -A because it respects the internal group paths of the source
echo "Step 2: Performing final merge..."
staged_files=(./tmp_groups/*.nc)

for tmp in "${staged_files[@]}"; do
    echo "Merging $tmp..."
    # --no_tmp_fl prevents creating a copy of test7.nc every time
    ncks -4 -A --no_tmp_fl "$tmp" test7_uncompressed.nc
done

# Compress and cleanup temporary files
echo 'Compressing and cleaning up'
ncks -4 -L 4 test7_uncompressed.nc test7.nc
rm test7_uncompressed.nc
rm -rf ./tmp_groups
echo "Done! Output saved to test7.nc"
```
